### ISSUE TITLE
Possible to have home-manager executable on PATH even when HM is used as a NixOS module?

### ISSUE THREAD

**Issue by alex-robbins** [nix-community/home-manager#3211]
(This might be more of a question than an issue if it turns out there's no bug; hopefully that's okay.) 

I'm using home-manager as a module of my NixOS system configuration flake, as in manual section ["3.3. NixOS module"]. I also wanted to have the `home-manager` executable on my PATH (see below for why), and was surprised to find that `programs.home-manager.enable = true` had no effect, although `home.packages = [ pkgs.home-manager ]` seems to work fine. 

Looking at the implementation, I see on [home-manager.nix:31] that `programs.home-manager.enable` explicitly has no effect when `config.submoduleSupport.enable` is true. 

- Is this check necessary? Is there some reason why (even when I've explicitly requested it) I should not be able to have `home-manager` on my PATH when using HM as a NixOS submodule?
- Other than evading that check, is there any practical difference between `programs.home-manager.enable = true` and `home.packages = [ pkgs.home-manager ]`? Any catch I'm not seeing?

[home-manager.nix:31]: https://github.com/nix-community/home-manager/blob/4cfc0a1e02c6374f66acdfd2ff8ae3e87c80c818/modules/programs/home-manager.nix#L31

["3.3. NixOS module"]: https://rycee.gitlab.io/home-manager/index.html#sec-flakes-nixos-module

---

For what it's worth, the reason I want to do this in the first place is because I'd normally like to use HM as a module in my system flake, but I'd also like to be able to use the same flake to build my HM generation separately from the system (in the standalone way). I expect to be able to do this by modifying my flake to provide a `homeConfigurations` attr in addition to the usual `nixosConfigurations`. This would be useful to iterate on the HM generation without unnecessarily rebuilding the system over and over. (If you have any thoughts about that or if I've missed something that makes this idea bad or impossible, I'm all ears.)

**Comment by ParetoOptimalDev** [comment]
This would be useful.

**Comment by rycee** [comment]
Yeah, I'd like to make that possible but we would first need to make the `home-manager` tool functionality dependent on whether it is installed from the NixOS module or from standalone installation. Specifically, we should disable the `switch` functionality when installed from the NixOS module.

That switching works even when installed through the NixOS module is a historical accident that I hope to resolve at some point. When installed through the NixOS module, we should include the HM generation directories inside the NixOS generation to improve reproducibility. When that happens it would no longer be possible to perform switching outside the regular NixOS flow, e.g., using `nixos-rebuild`.

**Comment by alex-robbins** [comment]
@rycee: I see how including the the HM generation inside the NixOS generation (when HM is used through the NixOS module) would be a good thing.

In the use case from my original comment, though, I think the end result that you're describing, where "it would no longer be possible to perform switching outside the regular NixOS flow" is not what I wanted. I was imagining that I could build and activate a NixOS configuration that includes an HM configuration for my user, and then subsequently build and switch to a standalone HM configuration for that same user, which would override the HM configuration from the system config.

Does that sound desirable to you? Also, if that's not what you were talking about, then what did you mean by "I'd like to make that possible"? I'm worried we might be talking about two different things.

Anyway, for my use case, maybe the answer is that I shouldn't be using HM as a NixOS module at all; maybe I should just use standalone HM, plus some custom scripts to boostrap HM on first login and keep it updated (since it would no longer be built with the system).

**Comment by panchoh** [comment]
Hi, folks.

I also have a config where I intend to use Home Manager both as a NixOS module and directly with the `home-manager` executable.

Until I found this issue report, what I observed was that running `nix-collect-garbage -d` would not purge the Home Manager generations. I would have to go and manually activate Home Manager, by running:

    nix run /etc/nixos#homeConfigurations.pancho.activationPackage

and then I would get back the `home-manager` executable available. By then running:

    home-manager generations
    
I could verify that they were a few old ones. I would then proceed to run:

        home-manager expire-generations '-1 sec'
        
to send them to oblivion.

But, after applying @alex-robbins's workaround (thanks!) of adding `home-manager` explicitly to `home.packages`, what I noticed was that then, running `nix-collect-garbage -d` would indeed delete Home Manager's generations! They could be seen in the output on the command, and running `home-manager generations` confirmed it.

I thought it relevant enough to share it here.

My config can be seen referenced in the line above this comment, if that's any help.

Thanks!

**Comment by packruler** [comment]
I am not sure if this helps when trying to intermingle `home-manager` executable with NixOS, but I came across this issue while trying to figure out a method for handling existing file conflicts.

My solution was to install the `home-manager` executable with in my flake configuration:
```nix
{
  inputs.home-manager.url = "github:nix-community/home-manager/release-23.05";

  outputs = { home-manager, ... }: rec {
    homeConfigurations.packruler = home-manager.lib.homeManagerConfiguration {
      pkgs = (import nixpkgs) {
        system = "x86_64-linux";
      };

      modules = [
        # import the same config that is used in NixOS config
        nixosModules.home
      ];
    };

    nixosModules = rec {
      home = {
        # install home-manager package from home-manager flake.
        home.packages = [ home-manager.packages.x86_64-linux.home-manager ];

        # ... all other home-manager configs
      };

      # nixosModules.default is used for building the home-manager config from NixOS
      default = {
        imports = [
          home-manager.nixosModules.home-manager
        ];

        home-manager.users.packruler = home;
      };
    };
  };
}
```
Now if I need to handle config conflicts I can run `home-manager switch --flake /<path_to_home_flake> -b backups`.

I have not really tested any other `home-manager` functionality, but I can confirm generation handling appears to be working properly.
